pyat/at/physics/frequency_maps.py

```diff
@@ -0,0 +1,283 @@
+
+    # returned array
+    xy_nuxy_lognudiff_array = numpy.empty([])
+    loss_map_array = numpy.empty([])
+
+    # verify steps
+    if numpy.count_nonzero(steps) != 2:
+        raise ValueError('steps can not be zero')
+
+    xmin = numpy.minimum(coords[0], coords[1])
+    xmax = numpy.maximum(coords[0], coords[1])
+    ymin = numpy.minimum(coords[2], coords[3])
+    ymax = numpy.maximum(coords[2], coords[3])
+    xsteps = steps[0]
+    ysteps = steps[1]
+
+    # define rectangle and x,y step size
+    if scale == "nonlinear":
+        verboseprint('non linear steps')
+        xinterval = 1.0*(xmax - xmin)
+        yinterval = 1.0*(ymax - ymin)
+        xauxsteps = numpy.arange(-0.5, 0.0 + 1e-9, 1./xsteps)
+        yauxsteps = numpy.arange(-0.5, 0.0 + 1e-9, 1./ysteps)
+        xnrsteps = 10**xauxsteps
+        xnrsteps = xnrsteps / sum(xnrsteps)
+        ynrsteps = 10**yauxsteps
+        ynrsteps = ynrsteps / sum(ynrsteps)
+        xscalesteps = numpy.concatenate((xnrsteps,
+                                        numpy.flip(xnrsteps)), axis=None)
+        yscalesteps = numpy.concatenate((ynrsteps,
+                                        numpy.flip(ynrsteps)), axis=None)
+        ixarray = xmin + 0.5 * xinterval * numpy.cumsum(xscalesteps)
+        iyarray = ymin + 0.5 * yinterval * numpy.cumsum(yscalesteps)
+    else:  # scale == 'linear', ignore any other
+        verboseprint('linear steps')
+        # get the intervals
+        xstep = 1.0*(xmax - xmin)/xsteps
+        ystep = 1.0*(ymax - ymin)/ysteps
+        ixarray = numpy.arange(xmin, xmax+1e-6, xstep)
+        iyarray = numpy.arange(ymin, ymax+1e-6, ystep)
+    lenixarray = len(ixarray)
+    leniyarray = len(iyarray)
+
+    print("Start tracking and frequency analysis")
+
+    # Forcing a maximum value of cpus
+    # I have no idea how to confirm the actual used value from patpass
+    # at.DConstant.patpass_poolsize = ncpu;
+    print(f' Requested POOL size : {ncpu}')
+
+    # tracking in parallel multiple x coordinates with the same y coordinate
+    for iy, iy_index in zip(iyarray, range(leniyarray)):
+        print(f'Tracked particles {abs(-100.0*iy_index/leniyarray):.1f} %',
+              end='')
+        print(f', with a max. cpu POOL size of {ncpu}')
+        verboseprint("y =", iy)
+        # z01 = numpy.array([ix*xscale+1e-9, 0,  0, 0,  0, 0])
+        # add 1 nm to tracking to avoid zeros in array for the ideal lattice
+        z0 = numpy.zeros((6, lenixarray))
+        # z0 = z0 + add_offset6D + addco[0,:];
+        # z0[0,:] = z0[0,:] + xscale*ixarray + 1e-9;
+        # z0[2,:] = z0[2,:] + yscale*iy      + 1e-9;
+        z0[0, :] = xscale * ixarray + add_offset6D[0] + addco[0][0] + 1e-9
+        z0[1, :] = add_offset6D[1] + addco[0][1]
+        z0[2, :] = yscale * iy + add_offset6D[2] + addco[0][2] + 1e-9
+        z0[3, :] = add_offset6D[3] + addco[0][3]
+        z0[4, :] = add_offset6D[4] + addco[0][4]
+        z0[5, :] = add_offset6D[5] + addco[0][5]
```

Comment:
The commented version is more efficient (with `z0 = z0 + add_offset6D + orbit`)
Is there any problem there ?

Reply:
@lfarv 
These changes have been implemented in commit https://github.com/atcollab/at/pull/556/commits/6503cf4a5baefdf74176a3feb77b4c5921160185

@lfarv thank you for the notes on the transpose and memory organization.
I implemented the changes and they work, but, I don't know the details of C-aligned and Fortran-aligned memory handling. Please, let me know if you have any documentation I could read on this topic.

Reply:
Ordering is specified in numpy by the "order" parameter in many function calls, for instance for[ array creation](https://numpy.org/doc/1.23/reference/generated/numpy.array.html#numpy-array). You can retrieve it in the ["flags" attribute](https://numpy.org/doc/1.23/reference/generated/numpy.ndarray.flags.html?highlight=ndarray%20flags#numpy-ndarray-flags) of any array.

More generally, [this article](https://en.wikipedia.org/wiki/Row-_and_column-major_order) describes the memory layout of multidimensional arrays.

Usually, you don't have to care about memory layout because numpy handles this for you. The problem appears when the data is processed in another language (C for AT): the C code must know how the data is organised in memory. So `lattice_pass` and the element constructors reorganise the arrays if necessary, but it' more efficient if you can avoid it.

For info, Fortran-style ordering was chosen because it's the way Matlab organises its data. numpy lets you chose, and the default is C-style.

Reply:
Thank you @lfarv